### PR TITLE
Houdini: Load Asset HDA in LOPs

### PR DESCRIPTION
**Pull request by BigRoy**
## Changelog Description

Implements a prototype for a loader HDA in LOPs.

- [x] Can be created via regular Houdini node workflows
- [x] Can be created via the Ayon loader
- [x] Can be managed via Ayon scene inventory
- [x] Allows to show the version thumbnail with the Node in the Network View using Network Editor background images
   - Only shows a thumbnail if the version has a thumbnail

To-do:
- [ ] Implement the "representation picker" UI - clicking the AYON icon on the node will now randomly pick a USD/ABC representation just for testing.
- [ ] Version dropdown menu on the node has not been implemented yet (it's just placeholder currently) - preferably it'd update the available entries as you click on it so you can pick a version from the currently existing ones.
- [ ] Code cleanup / todos
- [ ] Might be nice to see how we can adapt this to be able to re-use similar functionality in e.g. SOP imports, etc. but admittedly I'm not that much of an Houdini export so might need @fabiaserra or @MustafaJafar to help out with that. I would love to avoid having to manage 10s of HDAs since you can hardly bulk edit those for as far as I can see. (nor do they seem really version control-friendly)
- [ ] I believe currently on scene re-open all the logic that updates the representation id, etc. _might_ trigger again - if so, will need to clean that up so it only runs when the user is explicitly changing the values.

## Additional info

Some screen captures playing with the prototype:

https://github.com/ynput/OpenPype/assets/2439881/80258d1b-0767-4587-aef2-ae8354055ef7

https://github.com/ynput/OpenPype/assets/2439881/e31dc102-d747-4914-a2b5-beebf6eea0b5

https://github.com/ynput/OpenPype/assets/2439881/c85e5af5-580b-4e22-9cec-e255c2a179f5

Also shared [here](https://discord.com/channels/517362899170230292/517382145552154634/1166763783079874713) and [here](https://discord.com/channels/517362899170230292/517382145552154634/1167233290823008347).

**Feedback very welcome!**

## Testing notes:

1. Launch Houdini
2. Go to Solaris `/stage`
3. Start loading assets

### CODE REVIEW

**Comment by MustafaJafar** [comment]
I gave it a quick test as a user and managed to use it
I faced problems but with the publishing itself, not the loading. 
image

---

these are my note so far:
1. `File`: Houdini doesn't like backslashes, does path with backslashes work on your side ? 
2. value of `Primitive Root`: is that a convention on your side ? 
image

**Comment by BigRoy** [comment]
> 1. `File`: Houdini doesn't like backslashes, does path with backslashes work on your side ?

Same here - that's not due to the HDA by the way, the regular `Reference` node also seems to be uncapable of doing so.
A UNC path like your path does work with forward slashes.

> 2. value of `Primitive Root`: is that a convention on your side ?

Not at all - it was solely done for now to make it easy for testing, so that different loads would always get unique paths.

File: openpype/hosts/houdini/api/hda_utils.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        project_name = context["project_name"]
+        try:
+            repre_doc = get_representation_by_id(project_name, repre_id)
+        except Exception:
+            # Ignore invalid representation ids silently
+            repre_doc = None
+        update_info(node, repre_doc)
+
+        if repre_doc:
+            path = get_representation_path(repre_doc)
```
Comment (BigRoy) [inline openpype/hosts/houdini/api/hda_utils.py:131]:
```suggestion
            path = get_representation_path(repre_doc)
            # Load fails on UNC paths with backslashes and also 
            # fails to resolve @sourcename var with backslashed
            # paths correctly. So we force forward slashes
            path = path.replace("\\", "/")
```

**Comment by MustafaJafar** [comment]
This works so far, and I love it. 

--- 

I imagine the HDA with some buttons and drop down menus, e.g. a button select the asset, drop down menu to select the version. 
But, I think a button that shows a window that look exactly like the loader with no actions, we only double click and select. (is that what "representation picker" UI  means ?)

---

I could reuse some of the publisher code to select an asset an its tasks, but I'm not sure if it helps, anyways I'll keep digging. 

```python
"""Select an asset and get its tasks. """"

from openpype.pipeline import get_current_project_name
from openpype.client import get_asset_by_name

from openpype.tools.publisher.widgets.assets_widget import AssetsDialog
from openpype.tools.utils.host_tools import get_tool_by_name
from openpype.hosts.houdini.api.lib import get_main_window


project_name = get_current_project_name()

main_window = get_main_window()
publisher_window = get_tool_by_name( tool_name="publisher", parent=main_window)

dialog = AssetsDialog(publisher_window.controller, main_window)
dialog.exec_()

selected_asset = dialog.get_selected_asset()
if selected_asset:
    asset = get_asset_by_name(project_name, selected_asset)
    print ("asset name:", selected_asset)
    print("asset tasks:", ", ".join(asset["data"]["tasks"].keys()))
```
---

> I would love to avoid having to manage 10s of HDAs since you can hardly bulk edit those for as far as I can see.

Does adding functions definitions out side the HDAs helps , maybe in a `hda_lib.py ?

**Comment by BigRoy** [comment]
> This works so far, and I love it.

Anything in particular that's mostly lacking for you? Like - aside of the picker UI, anything that else you feel is still todo before a v1.0?

> I imagine the HDA with some buttons and drop down menus, e.g. a button select the asset, drop down menu to select the version. But, I think a button that shows a window that look exactly like the loader with no actions, we only double click and select. (is that what "representation picker" UI means ?)

The idea is that it'd show a UI that preferably isn't _massive_ but it'd essentially need to be a representation picker that allows you to pick:
- Project, asset, subset, version, representation.
- The 'version' I suppose could be left out and a "version" combobox attribute would then solve that aspect directly on the node.
- The 'representation' could just for the most part default to ".usd" for LOP imports, but I'm not sure if that's sufficient in the long run?

It's basically to large extent the **Loader** equivalent but then one that's hopefully simpler and is quicker to click through and select in. Another visual approach could be like the "Switch asset" dialog, since it allows you to quickly select asset, subset and representation with search filters.

The best case I imagine is clicking the button and seeing a dialog of "thumbnails" of available things to load and being able to easily search/filter by typing e.g. `villain` and seeing only the villain related usd data.

---

> > I would love to avoid having to manage 10s of HDAs since you can hardly bulk edit those for as far as I can see.
> 
> Does adding functions definitions out side the HDAs helps , maybe in a `hda_lib.py ?

This is already the case in `hda_utils.py` - however since certain aspects need to be in certain callbacks or settings of an HDA there's still quite some stuff to set in the different areas - however, my knowledge of managed HDAs is very limited so there might very well be decent ways to manage this nonetheless.

**Comment by MustafaJafar** [comment]
> Anything in particular that's mostly lacking for you? 

The only thing I can think of is "Should Ayon Houdini HDAs follow a particular convention? e.g. HDA naming, code logic location, etc.`

**Comment by BigRoy** [comment]
> > Anything in particular that's mostly lacking for you?
> 
> The only thing I can think of is "Should Ayon Houdini HDAs follow a particular convention? e.g. HDA naming, code logic location, etc.`

Good question - I'm not sure who's best to answer that.

**Comment by fabiaserra** [comment]
Our parm template for pointing to DB assets in Houdini looked something like this:

Screenshot 2023-11-20 at 5 03 46 PM

Of course the Episode/Sequence/Shot parms don't necessarily apply in OP/Ayon as it would simply be $AVALON_ASSET but those were super useful to filter out the consequent menus to whatever was selected (i.e., you wouldn't see all the Sequences of the project on the dropdown, only the ones for the selected Episode). 

This default parm was already very useful for example to set as a dummy importer on a wrapped Null node so artists could use it simply to channel reference the Path parameter for assets that didn't necessarily need to load anything on the scene.

As for naming convention on the HDA's I would probably stick to the defaults that Houdini uses when creating digital assets and you set an author and version, so it would be something like this (`<type>_<author>.<node_name>.<version>.hda`)
Screenshot 2023-11-20 at 5 11 52 PM

**Comment by mkolar** [comment]
@BigRoy this is coo, but I'm afraid it will need to be converted to be AYON compatible after the repo split we're working on. It's also been stuck in draft for a while, shall we close it and aim to reopen in ayon-houdini repo?

**Comment by BigRoy** [comment]
We have this particular one merged into our branch and have edits on top of it - we're using the HDA with some customizations.

I think it's ok to close this PR - but we'll need to definitely create a tracking issue on the separated repo pointing to this PR so we don't lose the information that this PR and its comments contains. Can you already cross-reference that as an issue there?

The work on this should at least continue - even if the endeavor would be Ayon focused.

**Comment by mkolar** [comment]
> We have this particular one merged into our branch and have edits on top of it - we're using the HDA with some customizations.
> 
> I think it's ok to close this PR - but we'll need to definitely create a tracking issue on the separated repo pointing to this PR so we don't lose the information that this PR and its comments contains. Can you already cross-reference that as an issue there?
> 
> The work on this should at least continue - even if the endeavor would be Ayon focused.

Everything we're closing is marked as `port to AYON` so we can keep track of all the PRs that were left out of OP. Next week we'll be creating starting repos for addons, so then we can put it there.

**Pull request closed without merging**